### PR TITLE
Add rest endpoint trigger_gc to trigger GC on Bookie

### PR DESCRIPTION
**Pull request by jiazhai**
Descriptions of the changes in this PR:

Add rest endpoint trigger_gc to trigger GC on Bookie

### Motivation

Some times user would like to trigger GC manually instead of waiting to the timeout or disk full. 

> ---
> In order to uphold a high standard for quality for code contributions, Apache BookKeeper runs various precommit
> checks for pull requests. A pull request can only be merged when it passes precommit checks. However running all
> the precommit checks can take a long time, some trivial changes don't need to run all the precommit checks. You
> can check following list to skip the tests that don't need to run for your pull request. Leave them unchecked if
> you are not sure, committers will help you:
>
> - [ ] [skip bookkeeper-server bookie tests]: skip testing `org.apache.bookkeeper.bookie` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server client tests]: skip testing `org.apache.bookkeeper.client` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server replication tests]: skip testing `org.apache.bookkeeper.replication` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server tls tests]: skip testing `org.apache.bookkeeper.tls` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server remaining tests]: skip testing all other tests in bookkeeper-server module.
> - [ ] [skip integration tests]: skip docker based integration tests. if you make java code changes, you shouldn't skip integration tests.
> - [ ] [skip build java8]: skip build on java8. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> - [ ] [skip build java9]: skip build on java9. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> ---

> ---
> Be sure to do all of the following to help us incorporate your contribution
> quickly and easily:
>
> If this PR is a BookKeeper Proposal (BP):
>
> - [ ] Make sure the PR title is formatted like:
>     `<BP-#>: Description of bookkeeper proposal`
>     `e.g. BP-1: 64 bits ledger is support`
> - [ ] Attach the master issue link in the description of this PR.
> - [ ] Attach the google doc link if the BP is written in Google Doc.
>
> Otherwise:
> 
> - [ ] Make sure the PR title is formatted like:
>     `<Issue #>: Description of pull request`
>     `e.g. Issue 123: Description ...`
> - [ ] Make sure tests pass via `mvn clean apache-rat:check install spotbugs:check`.
> - [ ] Replace `<Issue #>` in the title with the actual Issue number.
> 
> ---

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java [hunk 1]
```diff
@@ -382,6 +382,13 @@ public void runWithFlags(boolean force, boolean suspendMajor, boolean suspendMin
             lastMinorCompactionTime = System.currentTimeMillis();
             minorCompactionCounter.inc();
         }
+
+        if (force) {
+            if (forceGarbageCollection.compareAndSet(true, false)) {
+                LOG.info("{} Set forceGarbageCollection to true after force GC to make it forceGC-able again.", Thread
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: @jiazhai this is "set forcrGarbageCollection to false after force GC", no?

Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: Also doesn't https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java#L338 already do that?

File: bookkeeper-http/http-server/src/main/java/org/apache/bookkeeper/http/HttpRouter.java [hunk 2]
```diff
@@ -45,6 +45,7 @@
     public static final String LAST_LOG_MARK                = "/api/v1/bookie/last_log_mark";
     public static final String LIST_DISK_FILE               = "/api/v1/bookie/list_disk_file";
     public static final String EXPAND_STORAGE               = "/api/v1/bookie/expand_storage";
+    public static final String TRIGGER_GC                   = "/api/v1/bookie/trigger_gc";
```
Comment (sijie) [inline bookkeeper-http/http-server/src/main/java/org/apache/bookkeeper/http/HttpRouter.java]: I think it is better to just call it `/api/v1/bookie/gc`. a `PUT` method is to trigger gc and a `GET` method is to query the gc status. we can leave implementing `GET` in a separate pull request. but at least it allows extensibility.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedLedgerStorage.java [hunk 3]
```diff
@@ -191,6 +191,11 @@ public void diskJustWritable(File disk) {
         };
     }
 
+    @Override
+    public void forceGC() {
+        gcThread.enableForceGC();
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedLedgerStorage.java]: this is not actually implementing a "trigger" behavior. it only sets flag, however the actual gc will be happening after `gcWaitInterval`. so I think a better solution would be:

```suggestion
        gcThread.enableForceGC();
        // submit a task immediately
        gcExecutor.submit(this);
```

**Comment by sijie** [comment]
mark this for 4.7.3 as well, since this is needed for handling production issue introduced by #1837 without restarting bookies.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java (hunk 1 above)
Comment (jiazhai) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: Thanks. This depends. If `runWithFlags` is called in above safeRun(), this part of change is not needed.
In this PR it is force called by triggerGC, and will not use safeRun.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedLedgerStorage.java (hunk 3 above)
Comment (jiazhai) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedLedgerStorage.java]: Thanks, Maybe the original name is misleading. Here is what enableForceGC does in gcThread:
```
    public void enableForceGC() {
        if (forceGarbageCollection.compareAndSet(false, true)) {
            LOG.info("Forced garbage collection triggered by thread: {}", Thread.currentThread().getName());
            triggerGC(true, suspendMajorCompaction.get(),
                      suspendMinorCompaction.get());
        }
    }

    Future<?> triggerGC(final boolean force,
                        final boolean suspendMajor,
                        final boolean suspendMinor) {
        return gcExecutor.submit(() -> {
                runWithFlags(force, suspendMajor, suspendMinor);
            });
    }

```

Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedLedgerStorage.java]: i see

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java (hunk 1 above)
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java]: okay

**Comment by sijie** [comment]
@jiazhai can you fix the endpoint name? once it is fixed, we can move forward with 4.7.3 release.

**Pull request merged**